### PR TITLE
add support for arcadeshield to pxt-microbit

### PR DESCRIPTION
**Pull request by tballmsft**
## TODO (core)

- [x] change pxtarget.json to build both variants
- [x] review all "shim=" and "helper="
- [x] move pure STS code out to pxt-arcadeshield
- [x] make sure all shimmed functions have simulator support
- [x] test on hardware (MicroData)

## Blocks and Docs

- [x] move bitmap.d.ts and all other block defining stuff out to pxt-arcadeshield
- [x] all docs go to pxt-arcadeshield
- [x] remove "game" terminology
- [x] review the various blocks and APIs and add comments
   - [x] Bitmap (used to be Image) interface
   - [x] bitmap namespace
   - [x] screen (is a Bitmap)
   - [x] screenhelpers

### CODE REVIEW

**Comment by tballmsft** [comment]
some code

```
controller.A.onEvent(ControllerButtonEvent.Pressed, function () {
    if (controller.A.isPressed()) {
    	
    }
})
let ff = bmp`
    . . . . . . . . . . . . . . . . 
    . . . . . . . . . . . . . . . . 
    . . . . . . . . . . . . . . . . 
    . . . . . . . . . . . . . . . . 
    . . . . . 3 3 3 3 3 3 3 . . . . 
    . . . . 3 3 . . . . . . 3 . . . 
    . . . . . 3 . . . . . . 3 . . . 
    . . . . . 3 . . . . . . 3 . . . 
    . . . . . 3 . . . . . . 3 . . . 
    . . . . . 3 3 . . . . 3 . . . . 
    . . . . . . 3 3 3 3 3 3 . . . . 
    . . . . . . . . . . . . . . . . 
    . . . . . . . . . . . . . . . . 
    . . . . . . . . . . . . . . . . 
    . . . . . . . . . . . . . . . . 
    . . . . . . . . . . . . . . . . 
    `
ff.drawRect(0, 0, 0, 0, 0)
ff.setPixel(0, 0, 0)
let p = ff.getPixel(0, 0)
ff.fill(5)
ff.fillRect(0, 0, 1, 1, 6)
ff.drawLine(0, 0, 10, 10, 5)
ff.drawRect(2, 2, 5, 5, 5)
let c = ff.clone()
ff.flipX()
ff.flipY()
ff.replace(3, 4)
const b = ff.equals(c)
const w = ff.width
const h = ff.height
const b2 = ff.isMono
ff.copyFrom(c)
const buf = Buffer.create(100)
ff.getRows(3,buf)
ff.setRows(3,buf)
ff.transposed()
ff.scroll(2,2)
const dbl = ff.doubledX()
ff.doubledY()

```

File: sim/state/bitmap.ts [hunk 1]
```diff
@@ -0,0 +1,1135 @@
+namespace pxsim {
+    export class RefImage extends RefObject {
```
Comment (eanders-ms) [inline sim/state/bitmap.ts:2]: Suggestion: rename RefImage -> RefBitmap (unless you've already tried this and it didn't work for some reason). Avoiding duplicate global symbols is good practice, even if the `screen` and `bitmap` libs are never loaded together.

File: sim/state/bitmap.ts [hunk 2]
```diff
@@ -0,0 +1,1135 @@
...
+    export function doubledIcon(buf: RefBuffer): RefBuffer {
+        let img = ofBuffer(buf)
+        if (!img)
+            return null
+        img = BitmapMethods.doubled(img)
+        return toBuffer(img)
+    }
+}
+
+namespace pxsim.pxtcore {
```
Comment (eanders-ms) [inline sim/state/bitmap.ts:1113]: Everything in this namespace is more screen related than bitmap. To separate concerns, do you think it makes sense to move it into a separate file called `sim/arcadeshield.ts`?

File: libs/bitmap/bitmap.ts [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            let r = img.transposed();
+            r.flipX();
+            return r;
+        } else {
+            return null;
+        }
+    }
+
+    //% shim=pxt::setScreenBrightness
+    function _setScreenBrightness(brightness: number) { }
```
Comment (eanders-ms) [inline libs/bitmap/bitmap.ts:281]: Move brightness functions to arcadeshield?

File: sim/state/bitmap.ts (hunk 1 above)
Comment (tballmsft) [reply sim/state/bitmap.ts:2]: as discussed, can't do because of special support in compiler for RefImage

File: sim/state/bitmap.ts (hunk 2 above)
Comment (tballmsft) [reply sim/state/bitmap.ts:1113]: will do

File: libs/bitmap/bitmap.ts (hunk 3 above)
Comment (tballmsft) [reply libs/bitmap/bitmap.ts]: removed

**Review by eanders-ms — APPROVED** [review]

**Comment by eanders-ms** [comment]
"arcadeshield" and "bitmap" are showing up as available extensions. Should they be hidden?

image

**Pull request merged**